sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -31,17 +31,50 @@ import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy._
 
 trait DateTimeFormatterHelper {
+  private def getOrDefault(accessor: TemporalAccessor, field: ChronoField, default: Int): Int = {
+    if (accessor.isSupported(field)) {
+      accessor.get(field)
+    } else {
+      default
+    }
+  }
+
+  protected def toLocalDate(accessor: TemporalAccessor, allowMissingYear: Boolean): LocalDate = {
+    val year = if (accessor.isSupported(ChronoField.YEAR)) {
+      accessor.get(ChronoField.YEAR)
+    } else if (allowMissingYear) {
+      // To keep backward compatibility with Spark 2.x, we pick 1970 as the default value of year.
+      1970
+    } else {
+      throw new SparkUpgradeException("3.0",
+        "Year must be given in the date/timestamp string to be parsed. You can set " +
```

Comment:
Maybe also suggest the alternative workaround, e.g. prepending `'1970 ' to the string-to-be-parsed and prepending `'yyyy '` to the format string. That one works without the legacy setting so I'd say it's preferred.

Reply:
A problem is that, the pattern is flexible and the year may not be in the beginning, e.g. `MM/dd yyyy ...`. I don't think there is a workaround, unless we provide an API to allow users to set the default values.

Reply:
That's actually great -- if any part is missing from the format and the input, they can just provide a default by appending a constant to the input, _and_ a constant to the format string. I.e., the real format string is "MM dd" or "dd MM" or anything arbitrary but you're missing year, then you can *always* prepend "yyyy " to the format string, and "2009 " to the input value, and that fills out that part of the value.